### PR TITLE
Introduce consistent validation and error handling for money and date inputs

### LINKED ISSUE #38: Introduce consistent validation and error handling for money and date inputs
## Summary
Expense, income, and budget flows should validate money and date inputs consistently and return clear, predictable error responses when invalid data is submitted.

This issue improves request validation and error handling for the core financial flows so invalid inputs fail cleanly and expected client-side mistakes do not fall through to vague or generic failures.

## Why this matters
Even in a class project, a budgeting app should reject invalid financial data in a clear and consistent way. This improves product trust, simplifies debugging, and makes the app feel more complete.

## Scope
- Validate money and date inputs for expense, income, and budget flows
- Standardize invalid-input handling across the affected routes
- Return deterministic 4xx-style errors for malformed requests
- Reserve generic server failures for unexpected errors
- Update route tests to cover invalid payloads

## Acceptance Criteria
- Invalid money/date inputs are rejected consistently
- Expense, income, and budget routes handle malformed payloads predictably
- Tests cover common invalid request cases
- Error responses are clearer and more consistent than the current behavior

## Out of Scope
- Full frontend error-message redesign
- Session/auth lifecycle changes
- Major schema redesign

## Story Points
3

### PR DESCRIPTION
**Pull request by adillpickles**
## Description
This PR introduces more consistent validation and error handling for money and date inputs across the budget, expense, and income flows. The goal is to return clear, deterministic 4xx responses for invalid client input instead of relying on database failures or inconsistent route behavior.

This PR:
- adds a shared positive-money validation helper in the budget utilities
- applies deterministic 400 responses for invalid money/date inputs in expense create/update routes
- applies deterministic 400 responses for invalid money inputs in income create/update routes
- updates `/api/budget` validation to use the shared helper while preserving the category-budget behavior introduced by the newer budget model work
- adds and updates tests covering the current-state validation behavior

## Related User Story / Issue
Closes #38

## Type of Change
- [ ] New feature
- [x] Bug fix
- [x] Backend / API change
- [ ] UI change
- [ ] Database change
- [ ] Documentation
- [ ] Refactor

## How Has This Been Tested?
- [x] GitHub Actions / CI tests passed
- [x] Manual testing performed
- [ ] Not tested locally

### Test Notes
Verified locally with:
- `npm test -- --runInBand`
- `npm run build`

Also ran the focused route/helper test set during the validation pass:
- `npm test -- --runInBand budget.test.js expenses.test.js income.test.js`

## UI Changes (if applicable)
- [x] No UI changes
- [ ] UI updated (add screenshots if needed)

## Notes for Reviewers
This is a refreshed Issue #38 implementation rebased conceptually on top of:
- Issue #37 exact income date cleanup
- Issue #41 category budget / unified summary support

## Checklist
- [x] Linked to a user story or issue
- [X] Code builds / tests pass in CI
- [x] Ready for review

### CODE REVIEW

File: nextjs/__tests__/budget.test.js [hunk 1]
```diff
@@ -30,8 +31,17 @@ beforeEach(() => {
...
   authenticate.mockResolvedValue({ user: authorizedUser })
   budget.normalizeMonth.mockImplementation(value => value)
+  budget.isPositiveMoneyValue.mockImplementation((value) => {
+    if (typeof value === 'number') return Number.isFinite(value) && value > 0
+    if (typeof value !== 'string') return false
+    const trimmedValue = value.trim()
+    if (!trimmedValue) return false
+    const amount = Number(trimmedValue)
+    return Number.isFinite(amount) && amount > 0
+  })
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/__tests__/budget.test.js]: updated the tests to stop reimplementing the helper inline and to use the real `isPositiveMoneyValue` behavior instead.

File: nextjs/__tests__/budget.test.js [hunk 2]
```diff
@@ -255,6 +276,22 @@ describe('normalizeDate', () => {
...
+    expect(actualBudget.isPositiveMoneyValue('25.50')).toBe(true)
+    expect(actualBudget.isPositiveMoneyValue(' 25.50 ')).toBe(true)
+  })
+
+  it('rejects empty, non-numeric, and non-positive values', () => {
+    expect(actualBudget.isPositiveMoneyValue('')).toBe(false)
+    expect(actualBudget.isPositiveMoneyValue('abc')).toBe(false)
+    expect(actualBudget.isPositiveMoneyValue(0)).toBe(false)
+    expect(actualBudget.isPositiveMoneyValue('-5')).toBe(false)
+    expect(actualBudget.isPositiveMoneyValue(null)).toBe(false)
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/__tests__/budget.test.js:674]: added coverage for non storable positive values so the helper is tested against DB relevant edge cases too.

File: nextjs/src/lib/budget.js [hunk 3]
```diff
@@ -58,6 +58,20 @@ export function normalizeDate(value) {
...
+  }
+
+  if (typeof value !== 'string') return false
+
+  const trimmedValue = value.trim()
+  if (!trimmedValue) return false
+
+  const amount = Number(trimmedValue)
+  return Number.isFinite(amount) && amount > 0
+}
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/lib/budget.js:86]: tightened `isPositiveMoneyValue` so it only accepts positive money values that match the app’s current storage expectations

File: nextjs/__tests__/expenses.test.js [hunk 4]
```diff
@@ -18,8 +22,19 @@ beforeEach(() => {
...
   authenticate.mockResolvedValue({ user: { id: 'uid', email: 'a@b.com' } })
   evaluateThresholdForMonth.mockResolvedValue(null)
+  isPositiveMoneyValue.mockImplementation((value) => {
+    if (typeof value === 'number') return Number.isFinite(value) && value > 0
+    if (typeof value !== 'string') return false
+    const trimmedValue = value.trim()
+    if (!trimmedValue) return false
+    const amount = Number(trimmedValue)
+    return Number.isFinite(amount) && amount > 0
+  })
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/__tests__/expenses.test.js]: updated the tests to stop reimplementing the helper inline and to use the real `isPositiveMoneyValue` behavior instead

File: nextjs/__tests__/income.test.js [hunk 5]
```diff
@@ -19,8 +19,17 @@ beforeEach(() => {
...
   authenticate.mockResolvedValue({ user: { id: 'uid', email: 'a@b.com' } })
   normalizeDate.mockImplementation((value) => value)
+  isPositiveMoneyValue.mockImplementation((value) => {
+    if (typeof value === 'number') return Number.isFinite(value) && value > 0
+    if (typeof value !== 'string') return false
+    const trimmedValue = value.trim()
+    if (!trimmedValue) return false
+    const amount = Number(trimmedValue)
+    return Number.isFinite(amount) && amount > 0
+  })
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/__tests__/income.test.js]: updated the tests to stop reimplementing the helper inline and to use the real `isPositiveMoneyValue` behavior instead

File: nextjs/src/lib/budget.js [hunk 6]
```diff
@@ -58,6 +60,22 @@ export function normalizeDate(value) {
   return getNormalizedDateString(value)
 }
 
+export function isPositiveMoneyValue(value) {
+  if (typeof value === 'number') {
+    if (!Number.isFinite(value) || value <= 0 || value > MAX_MONEY_VALUE) return false
+    return Number(value.toFixed(2)) === value
+  }
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/lib/budget.js:76]: tightened the helper without using a strict floating-point equality check, so valid two decimal numeric inputs with normal JS precision artifacts are still accepted

File: nextjs/__tests__/budget.test.js [hunk 7]
```diff
@@ -255,6 +270,31 @@ describe('normalizeDate', () => {
   })
 })
 
+describe('isPositiveMoneyValue', () => {
+  it('accepts positive numbers and numeric strings', () => {
+    expect(actualBudget.isPositiveMoneyValue(25)).toBe(true)
+    expect(actualBudget.isPositiveMoneyValue('25.50')).toBe(true)
+    expect(actualBudget.isPositiveMoneyValue(' 25.50 ')).toBe(true)
+  })
+
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/__tests__/budget.test.js:663]: added regression coverage for numeric fp artifact cases so valid 2 decimal amounts don’t get rejected because of JS precision noise

File: nextjs/src/lib/budget.js [hunk 8]
```diff
@@ -58,6 +61,25 @@ export function normalizeDate(value) {
   return getNormalizedDateString(value)
 }
 
+export function isPositiveMoneyValue(value) {
+  if (typeof value === 'number') {
+    if (!Number.isFinite(value) || value <= 0 || value > MAX_MONEY_VALUE) return false
+
+    const cents = value * 100
+    const roundedCents = Math.round(cents)
+    return Math.abs(cents - roundedCents) <= MONEY_EPSILON
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/lib/budget.js]: good catch

[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/lib/budget.js]: tightened the cents check so values must resolve to at least 1 cent, preventing tiny positives from slipping through

File: nextjs/src/app/api/income/update/route.js [hunk 9]
```diff
@@ -11,6 +11,10 @@ export async function POST(request) {
   const { income_id, source_id, amount, date, notes } = body
   if (!income_id) return NextResponse.json({ error: 'income_id required' }, { status: 400 })
 
+  if (amount !== undefined && !isPositiveMoneyValue(amount)) {
+    return NextResponse.json({ error: 'amount must be greater than 0' }, { status: 400 })
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/app/api/income/update/route.js]: Updated the message to reflect the full validation rule for valid positive money amounts

File: nextjs/src/app/api/income/route.js [hunk 10]
```diff
@@ -27,7 +27,10 @@ export async function POST(request) {
   let body = {}
   try { body = await request.json() } catch {}
   const { source_id, amount, date, notes } = body
-  if (!amount || !date) return NextResponse.json({ error: 'amount and date are required' }, { status: 400 })
+  if (amount == null || date == null) return NextResponse.json({ error: 'amount and date are required' }, { status: 400 })
+  if (!isPositiveMoneyValue(amount)) {
+    return NextResponse.json({ error: 'amount must be greater than 0' }, { status: 400 })
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/app/api/income/route.js]: Updated the message to reflect the full validation rule instead of only the > 0 condition

File: nextjs/src/app/api/expenses/update/route.js [hunk 11]
```diff
@@ -11,7 +11,21 @@ export async function POST(request) {
   const { expense_id, category_id, amount, description, date } = body
   if (!expense_id) return NextResponse.json({ error: 'expense_id required' }, { status: 400 })
 
-  const entries = Object.entries({ category_id, amount, description, date }).filter(([, v]) => v !== undefined)
+  if (amount !== undefined && !isPositiveMoneyValue(amount)) {
+    return NextResponse.json({ error: 'amount must be greater than 0' }, { status: 400 })
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/app/api/expenses/update/route.js]: Adjusted the error message to match the actual shared money validation rule

File: nextjs/src/app/api/expenses/route.js [hunk 12]
```diff
@@ -27,12 +27,21 @@ export async function POST(request) {
   let body = {}
   try { body = await request.json() } catch {}
   const { category_id, amount, description, date } = body
-  if (!amount || !date) return NextResponse.json({ error: 'amount and date are required' }, { status: 400 })
+  if (amount == null || date == null) {
+    return NextResponse.json({ error: 'amount and date are required' }, { status: 400 })
+  }
+  if (!isPositiveMoneyValue(amount)) {
+    return NextResponse.json({ error: 'amount must be greater than 0' }, { status: 400 })
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/app/api/expenses/route.js]: Updated the wording so it aligns with the helper’s full money validation rule

File: nextjs/src/app/api/budget/route.js [hunk 13]
```diff
@@ -28,7 +28,7 @@ export async function POST(request) {
   const monthlyLimit = body.monthly_limit
 
   if (!month) return NextResponse.json({ error: 'Valid month is required' }, { status: 400 })
-  if (monthlyLimit == null || Number.isNaN(Number(monthlyLimit)) || Number(monthlyLimit) <= 0) {
+  if (!isPositiveMoneyValue(monthlyLimit)) {
     return NextResponse.json({ error: 'monthly_limit must be greater than 0' }, { status: 400 })
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/src/app/api/budget/route.js]: Updated the message so it reflects the full validation rule rather than only the > 0 requirement

File: nextjs/__tests__/income.test.js [hunk 14]
```diff
@@ -268,6 +302,30 @@ describe('POST /api/income/update', () => {
...
   })
+
+  it('rejects update with a non-positive amount', async () => {
+    await testApiHandler({
+      appHandler: updateHandler,
+      async test({ fetch }) {
+        const res = await fetch(post({ income_id: 1, amount: 'abc' }))
+        expect(res.status).toBe(400)
+        expect((await res.json()).error).toBe('amount must be a valid positive money amount')
+      }
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/__tests__/income.test.js]: I renamed the test to match the actual validation being covered.

File: nextjs/__tests__/expenses.test.js [hunk 15]
```diff
@@ -207,6 +280,43 @@ describe('POST /api/expenses/update', () => {
...
     })
   })
+
+  it('400 - rejects a non-positive update amount', async () => {
+    await testApiHandler({
+      appHandler: updateHandler,
+      async test({ fetch }) {
+        const res = await fetch(post({ expense_id: 1, amount: 'abc' }))
+        expect(res.status).toBe(400)
+        expect((await res.json()).error).toBe('amount must be a valid positive money amount')
```
[the remark this answers is not in the document]
Comment (adillpickles) [reply nextjs/__tests__/expenses.test.js]: updated the title for clarity

**Review by zakariac22 — APPROVED** [review]
